arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,315 @@
+                     "joint edge images comprised of both horizontal and "
+                     "vertical information." );
+  config->set_value( "smoothing_sigma",
+                     d->smoothing_sigma,
+                     "Smoothing sigma for the output NMS edge density map." );
+  config->set_value( "smoothing_half_step",
+                     d->smoothing_half_step,
+                     "Smoothing half step for the output NMS edge density "
+                     "map." );
+
+  return config;
+}
+
+// ----------------------------------------------------------------------------
+void
+aligned_edge_detection
+::set_configuration( vital::config_block_sptr in_config )
+{
+  // Starting with our generated vital::config_block to ensure that assumed
+  // values are present. An alternative is to check for key presence before
+  // performing a get_value() call.
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config( in_config );
+
+  // Settings for edge detection
+  d->threshold = config->get_value< float >( "threshold" );
+  d->produce_joint_output =
+    config->get_value< bool >( "produce_joint_output" );
+  d->smoothing_sigma = config->get_value< double >( "smoothing_sigma" );
+  d->smoothing_half_step =
+    config->get_value< unsigned >( "smoothing_half_step" );
+}
+
+// ----------------------------------------------------------------------------
+bool
+aligned_edge_detection
+::check_configuration( vital::config_block_sptr config ) const
+{
+  return true;
+}
+
+// ----------------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+aligned_edge_detection
+::filter( kwiver::vital::image_container_sptr source_image_ptr )
+{
+  if( !source_image_ptr )
+  {
+    LOG_ERROR( logger(), "Image pointer was not valid." );
+    return nullptr;
+  }
+
+  // Get input image
+  vil_image_view_base_sptr source_image =
+    vxl::image_container::vital_to_vxl( source_image_ptr->get_image() );
+
+  // Perform Basic Validation
+  if( !source_image )
+  {
+    LOG_ERROR( logger(), "Image was not valid." );
+    return nullptr;
+  }
+
+  // Perform Basic Validation
+  if( source_image->nplanes() == 3 )
+  {
+    source_image = vil_convert_to_grey_using_average( source_image );
+  }
+  else if( source_image->nplanes() != 1 )
+  {
+    LOG_ERROR( logger(), "Input must be a grayscale image!" );
+    return nullptr;
+  }
+
+#define HANDLE_CASE( T )                                          \
+  case T:                                                         \
+  {                                                               \
+    using ipix_t = vil_pixel_format_type_of< T >::component_type; \
+    auto filtered = d->filter< ipix_t >( source_image );          \
+    auto container = vxl::image_container{ filtered };            \
+    return std::make_shared< vxl::image_container >( container ); \
+  }                                                               \
```

Comment:
You should be able to avoid a copy here:
```suggestion
    return std::make_shared< vxl::image_container >( filtered );  \
  }
```

(I also recommend omitting the `\` from the last line of a `#define`.)